util/gem5_bridge/src/gem5_bridge.c

```diff
@@ -0,0 +1,344 @@
+        return -EINVAL;
+    }
+
+    pr_info("%s: SUCCESS!\n", __func__);
+    return 0;
+}
+
+static ssize_t gem5_bridge_read(struct file *filp, char *ubuff,
+                                    size_t len, loff_t *offp)
+{
+    ssize_t result;
+    char *kbuff;
+    const char *file_name = filp->f_path.dentry->d_iname;
+    struct gem5_op *op = gem5_ops_search(file_name);
+
+    /* Allocate kernel buffer */
+    kbuff = kvmalloc(len, GFP_KERNEL);
+    if (!kbuff) {
+        pr_err("%s: unable to allocate kernel buffer\n", __func__);
+        return -ENOMEM;
+    }
+
+    /* Check that the device is valid and its op has read functionality */
+    if (!op || !op->read) {
+        pr_err("%s: read unsupported for "PATH"%s\n", __func__, op->name);
+        kvfree(kbuff);
+        return -ENOTTY;
+    }
+
+    /* Run this op's read function */
+    result = op->read(op, kbuff, len, offp);
+    if (result < 0) {
+        pr_err("%s: read failed for "PATH"%s\n", __func__, op->name);
+        kvfree(kbuff);
+        return result; /* propogate the error */
+    }
+
+    /* Copy kernel buffer to user buffer */
+    if (copy_to_user(ubuff, kbuff, result) != 0) {
+        pr_err("%s: unable to copy kernel buffer to user buffer\n", __func__);
+        kvfree(kbuff);
+        return -EFAULT;
+    }
+
+    pr_info("%s: SUCCESS! (result=%lu)\n", __func__, result);
+    kvfree(kbuff);
+    return result;
+}
+
+static ssize_t gem5_bridge_write(struct file *filp, const char *ubuff,
+                                    size_t len, loff_t *offp)
+{
+    ssize_t result;
+    char *kbuff;
+    const char *file_name = filp->f_path.dentry->d_iname;
+    struct gem5_op *op = gem5_ops_search(file_name);
+
+    /* Allocate kernel buffer */
+    kbuff = kvmalloc(len, GFP_KERNEL);
+    if (!kbuff) {
+        pr_err("%s: unable to allocate kernel buffer\n", __func__);
+        return -ENOMEM;
+    }
+
+    /* Copy user buffer to kernel buffer */
+    if (copy_from_user(kbuff, ubuff, len) != 0) {
+        pr_err("%s: unable to copy user buffer to kernel buffer\n", __func__);
+        kvfree(kbuff);
+        return -EFAULT;
+    }
+
+    /* Check that the device is valid and its op has write functionality */
+    if (!op || !op->write) {
+        pr_err("%s: write unsupported for "PATH"%s\n", __func__, op->name);
+        kvfree(kbuff);
+        return -ENOTTY;
+    }
+
+    /* Run this op's write function */
+    result = op->write(op, kbuff, len, offp);
+    if (result < 0) {
+        pr_err("%s: write failed for "PATH"%s\n", __func__, op->name);
+        kvfree(kbuff);
+        return result; /* propogate the error */
+    }
+
+    pr_info("%s: SUCCESS! (result=%lu)\n", __func__, result);
+    kvfree(kbuff);
+    return result;
+}
+
+static struct file_operations gem5_bridge_fops = {
+    .owner      = THIS_MODULE,
+    .open       = gem5_bridge_open,
+    .release    = gem5_bridge_release,
+    .mmap       = gem5_bridge_mmap,
+    .read       = gem5_bridge_read,
```

Comment:
I would strongly suggest that we don't support read and write. It doesn't really make sense to use this interface since it forces a transition into the kernel which seems undesirable. It also adds unnecessary (including some pretty nasty inline assembly) complexity in the driver. The only reason I can see for this interface is to enable call directly from shell scripts that write to the individual op files, but that doesn't seem like justification enough to replicate this part of the m5 tool.

Reply:
While I agree with the sentiment, I feel that it is important to keep these read/write implementations here to have functionality that matches the current set of operations. The m5 utility will remain, especially for the purposes of linking against it to truly prevent overhead during workloads, but the driver will be a useful platform for moving forward with some of our ideas for new gem5ops interfaces without dealing with the complexity of the m5 utility -- which, despite the complexity here, is far more daunting to modify. Because of this, it is useful to exemplify how to replicate these other ops in the driver, so that future changes will have a basis to work from.

As far as the inline assembly, it is unavoidable even in the m5 utility with the current interface. Honestly, that is part of the reason for progressing the bridge interface as the current method of abusing the calling convention to pass all of the arguments is a bit untenable.

Reply:
If the M5 utility remains for low-overhead operation, what's the point of the read/write endpoints? With very few exceptions, I can't see how a transition to kernel space would have an acceptably low impact on simulation results for most calls. Effectively, we just end up with two implementations of the same thing. If you really want this functionality, is there any reason you can't just link with `libgem5` to avoid code duplication?

I'm also worried that putting a lot of complexity in the kernel driver is going to be a maintenance nightmare, especially since this will never be upstream. Historically, all of the gem5-specific kernel extensions have been poorly maintained. My feeling is that the more complexity we add here, the more likely it is that this module will end up with maintenance issues (more complexity means a higher likelihood that it won't compile on older / newer kernels).

Reply:
Ok yeah, after discussing I think you're correct that, for an initial implementation, it should be as simple as possible until we know we need any other features. I am moving the prior code into a draft/reference PR here https://github.com/gem5/gem5/pull/1555 so that we still have the driver interface to experiment with.